### DISCUSSION TITLE
Authentifizierung auf der Zentrale aktivieren, ist das wirklich notwendig?

### DISCUSSION THREAD

**Discussion by Baxxy13** [SukramJ/aiohomematic#1119] [Q&A] [answered]
Da mir das immer wieder über den Weg läuft (im HomeMatic-Forum, bei Diskussionen und Tickets) und auch so in der [ReadMe](https://github.com/danielperna84/custom_homematic/blob/devel/README.md#authentication) steht...

> This integration always passes credentials to the HomeMatic hub when connecting. For CCU and descendants (RaspberryMatic, debmatic) **it is required** to enable authentication for XmlRPC communication (Settings/Control panel/Security/Authentication). JsonRPC communication ia always authenticated.

...muss ich mal nachhaken. 

Warum muss die Authentifizierung aktiviert sein?

Grund der Frage...
Bei mir laufen 3 SetUps (3x HA samt Integration mit jeweils 1x verbundene RM) ohne aktivierte Authentifizierung völlig problemlos.

**Comment by SukramJ** [comment]
Gute Frage. Ich hab das noch nicht systematisch untersucht.
Für die JSON-Abfragen benötigen wir die Benutzerauthentifizierung.
Für XmlRPC verwenden wir die gleichen Benutzeranmeldedaten.
Bei vielen Benutzer führt die Aktivierung von Authentifizierung zum Erfolg (oder ist das nur ein Nebeneffekt?).
Wenn ich die Authentifizierung in der CCU wieder deaktiviere habe ich auch keine Probleme.
Aber vielleicht hat @danielperna84 da noch mehr Infos.

**Comment by Baxxy13** [comment]
Hmm, die JSON-RPC Abfragen richten sich ja nach dem USER-Level. 
z.B: 
```
SysVar.getAll {
  LEVEL USER
  SCRIPT_FILE sysvar/getall.tcl
  INFO {Liefert Detailinformationen zu allen Systemvariablen auf der HomeMatic Zentrale}
  ARGUMENTS {_session_id_}
}
```
Hier muss also zwingend USER/PASS und Session-ID übergeben werden, aber die Authentifizierung "aktiv/inaktiv" sollte keine Rolle spielen. 

Das man hierfür in den Config-Flow einen "Admin" einträgt ist nachvollziehbar um Zugriff auf alle Methoden zu haben.

Bei der XmlRPC - Api habe ich gar keinen Plan. 
Wenn ich den Hilfetext richtig deute wird USER/PASS aber nur benötigt wenn die Checkbox (Authentifizierung:) aktiv ist.

Ich schreib mir das mal auf den Merkzettel und werde das bei Gelegenheit mal durchtesten...

Aber vielleicht wissen ja @danielperna84 oder @jens-maus mehr.

**Reply by SukramJ** [reply]
Wenn ich das mal in der README nachvollziehe, dann war Authentifizierung Anfangs nicht erwähnt, dann empfohlen, und später Pflicht.
Diese Eskalation habe, wenn ich die History so nachvollziehe, wohl ich betrieben, wobei mir die Gründe aktuell nicht mehr klar sind.
Ich denke das war jeweils die Reaktion auf Issues.

**Reply by jens-maus** [reply]
Nun, für alle die JSON-RPC Methoden die ja über den normalen Webserver der CCU/RaspberryMatic laufen muss man hier wie von @Baxxy13 korrekt gezeigt einen User+Passwort angeben der über Admin-Rechte verfügt. Alles andere ist/wird nicht zielführend sein. Deshalb muss man ja auch in der Homematic(IP) Local integration zwingend den CCU-Nutzername+CCU-Passwort angeben und kommt nicht drumherum. 

Bei XMLRPC Und der ReGaHss Skript-Kommunikation verhält sich das jedoch anders. Dort ist es abhängig von der Einstellungen in der WebUI (Systemsteuerung->Sicherheit->Authentifizierung) ob dort eine zwingende Authentifizierung gegenüber einem validen Admin-User auf der CCU passiert oder nicht.

D.h. also für JSON-RPC Abfragen braucht man IMMER (egal ob Systemsteuerung->Sicherheit->Authentifizierung an/aus ist) einen Admin-Useraccount den man in die Integration einträgt. Und Wenn man XMLRPC oder ReGa-Skript Abfragen macht kommt es darauf an was der Nutzer in der WebUI an/ausgeschalten hat ob eine Authentifizierung erzwungen wird oder nicht. Wie die Integration das jedoch intern managed kann ich natürlich wegen zu wenig einblick nicht sagen, das müsste @SukramJ mal prüfen wie da der Ablauf ist. Ggf. gibt es da schon eine Prüfung ob XMLRPC/ReGaHss ein Username/Passwort benötigen oder nicht und dann werden einfach die selben Credentials wie für das JSONRPC geliefert.

**Reply by SukramJ** [reply]
Also die Integration prüft nicht ob die Authentifizierung in der CCU aktiv ist. Die Credentials werden immer mit gesendet.

**Reply by jens-maus** [reply]
Dann würde mich dazu mal interessieren wie da genau der Payload dazu aussieht bzw. wie da genau der Ablauf ist. Vllt. kannst du dazu ja mal ein Beispiel liefern bzw. Links hierher packen wo man das nachvollziehen kann.

**Reply by SukramJ** [reply]
Im Header werden Benutzername/Passwort Base64 kodiert [abgelegt](https://github.com/danielperna84/hahomematic/blob/502f9cc5aef837cd628a8733b13a14e4fd8cd52e/hahomematic/support.py#L67-L74) und beim [Proxy](https://github.com/danielperna84/hahomematic/blob/502f9cc5aef837cd628a8733b13a14e4fd8cd52e/hahomematic/xml_rpc_proxy.py#L79C15-L79C30)  in den args mitgegeben.

Was ich bisher nicht nachvollzogen habe ist wie sich eine "frische" CCU/RM verhält.

**Reply by jens-maus** [reply]
Ok, dann fügt ihr anscheinend immer einen `Authorization:` header hinzu egal ob die XMLRPC Authentifizierung in der WebUI der CCU angeschalten ist oder nicht. Im Falle das sie dort deaktiviert ist wird das halt entsprechend ignoriert. Ist zwar prinzipiell kein Problem, aber dann drängt sich schon die Frage auf ob man dann nicht vielleicht unnötigerweise in diesem Fall den Nutzernamen+Passwort im Netzwerktraffic preis gibt (gerade wenn keine HTTPS/TLS Verschlüsselung genutzt wird). Besser wäre es natürlich diesen `Authorization:` header nur dann zu senden wenn die XMLRPC Authentifizierung wirklich notwendig ist.

**Reply by SukramJ** [reply]
Mal schauen was @Baxxy13 so rausfindet, dann werde ich da vielleicht noch mal nacharbeiten.

**Comment by Baxxy13** [comment]
Sehr gut, danke. Das deckt sich soweit mit meinen eigenen Erfahrungen.

Ich halte für mich fest das definitiv ein User mit Admin-Level in den Config-Flow der Integration eingetragen werden muss und die (aktivierte/deaktivierte) Authentifizierung auf der Zentrale damit quasi keine Rolle mehr spielt da ja die Credentials immer mit übergeben werden.

Soweit ich das beurteilen kann "ignoriert" z.B. die Remote-Script-Api die mitgesendeten Credentials wenn die Authentifizierung deaktiviert ist. Wie sich die XmlRPC-Api dahingehend verhält kann ich aktuell nicht sagen, denke aber genauso.
Von daher ist meine Frage beantwortet. 
"Authentifizierung aktiv" ist nicht nötig.

**Reply by SukramJ** [reply]
Die Frage, die dann bleibt ist, woran es in #1109 ansonsten gehapert hat.

**Reply by Baxxy13** [reply]
Ich hatte das verfolgt. Das war auch einer der Gründe für das "nachhaken" hier.
Erklären kann ich es auch nicht. 
Aber wie geschrieben teste ich das bei Gelegenheit mal mit komplett "jungfräulichen" Systemen durch.

**Reply by SukramJ** [reply]
Das wäre super. Mir fehlte dazu bisher Zeit und Motivation.

**Reply by Baxxy13** [reply]
Kein Problem. Hier langweilte sich gerade ein Pi4B mit frischem HA-OS und eine RM-VM mit 2 angelernten Aktoren...

Also die Integration "from Scratch" aufgesetzt und nur den Admin-User samt Passwort eingetragen.
Auf der RM ist die Authentifizierung NICHT aktiv, Firewall: Port's offen / Vollzugriff in allen Kategorien.

Es wurden alle Geräte, Programme und SysVars gefunden, keine Fehlermeldungen in HA. In der RM nur ein Eintrag der m.E. aber vernachlässigbar ist.
```
Jul 24 20:25:49 homematic-raspi user.err rfd: XmlRpcClient error calling newDevices({"RaspberryMatic_VM_96-BidCos-RF",{[ADDRESS:"BidCoS-RF:16",AES_ACTIVE:0,DIRECTION:1,FLAGS:1,INDEX:16,LINK_SOURCE_ROLES:"KEYMATIC REMOTECONTROL_RECEIVER SWITCH WINMATIC",LINK_TARGET_ROLES:"",PARAMSETS:{"LINK","MASTER","VALUES"},PARENT:"BidCoS-RF",PARENT_TYPE:"HM-RCV-50",TYPE:"VIRTUAL_KEY",VERSION:6],[ADDRESS:"BidCoS-RF:17",AES_ACTIVE:0,DIRECTION:1,FLAGS:1,INDEX:17,LINK_SOURCE_ROLES:"KEYMATIC REMOTECONTROL_RECEIVER SWITCH WI
MATIC",LINK_TARGET_ROLES:"",PARAMSETS:{"LINK","MASTER","VALUES"},PARENT:"BidCoS-RF",PARENT_TYPE:"HM-RCV-50",TYPE:"VIRTUAL_KEY",VERSION:6],[ADDRESS:"BidCoS-RF:18",AES_ACTIVE:0,DIRECTION:1,FLAGS:1,INDEX:18,LINK_SOURCE_ROLES:"KEYMATIC REMOTECONTROL_RECEIVER SWITCH WINMATIC",LINK_TARGET_ROLES:"",PARAMSETS:{"LINK","MASTER","VALUES"},PARENT:"BidCoS-RF",PARENT_TYPE:"HM-RCV-50",TYPE:"VIRTUAL_KEY",VERSION:6],[ADDRESS:"BidCoS-RF:19",AES_ACTIVE:0,DIRECTION:1,FLAGS:1,INDEX:19,LINK_SOURCE_ROLES:"KEYMATIC REMOTECONTROL
RECEIVER SWITCH WINMAT
```
Ich kann damit auch nichts anfangen und das soll hier auch nicht Thema sein.

Damit bestätigt sich für mich:
> das definitiv ein User mit Admin-Level in den Config-Flow der Integration eingetragen werden muss und die (aktivierte/deaktivierte) Authentifizierung auf der Zentrale damit quasi keine Rolle mehr spielt

**Reply by jens-maus** [reply]
... ausser wie gesagt das es IMHO besser wäre im Falle einer nicht notwendigen Authentifizierung den Auth header aus dem web request wegzulassen damit dieser nicht abgegriffen und damit die credentials abgegriffen werden können.

**Reply by Baxxy13** [reply]
Gibt es denn eine Methode um zu erkennen ob die Authentifizierung aktiv ist oder nicht? 
Wenn nicht, ist und bleibt der nutzerfreundlichste Weg, den Authentification-Header immer mitzusenden.
So kann der User quasi nix falsch machen.
 
Im lokalen LAN oder auch per VPN halte ich den lesbaren Header für vernachlässigbar. 
Wobei VPN wohl auch die wenigsten machen da der Datendurchsatz recht hoch ist. Wenn man da keine Flat hat ist ruck zuck das Datenvolumen weg.

**Reply by SukramJ** [reply]
> Gibt es denn eine Methode um zu erkennen ob die Authentifizierung aktiv ist oder nicht?

In der JsonAPI: `CCU.getAuthEnabled`

**Reply by Baxxy13** [reply]
Lässt sich das nutzen ohne den Config-Flow zu "verkomplizieren"?

Von meiner Seite sind die Fakten soweit klar. 
Das ein User mit Admin Rechten benötigt wird steht ja so schon in der ReadMe. 
Bei der Authentifizierung würde ich das "it is required" in "it is recommended" ändern.

**Reply by SukramJ** [reply]
> Lässt sich das nutzen ohne den Config-Flow zu "verkomplizieren"?

Mal schauen ...


Vielen Dank fürs Testen.

**Reply by SukramJ** [reply, marked as answer]
Für den ersten Aufruf (getVersion, um das Backend zu ermitteln) benötige ich noch die Authentifizierung.
Alle Folgeaufrufe senden dann ggf. keinen Authorization header mehr mit.
Das kann mit der Beta 1.40.0b0 getestet werden. Diese Version benötigt allerdings auch die HA Beta 2023.8.0b0.

**Comment by danielperna84** [comment]
Der Vollständigkeit halber: In der alten Integration waren die Credentials __nur__ für JSON-RPC notwendig. Also für die Namensauflösung. Wenn da jemand was falsches drin hatte, hat man dann die Entities ohne schöne Namen bekommen. Dadurch ist einigen oftmals nicht aufgefallen, dass es ein Problem mit der Authentifizierung gab, weil die dachten "das soll so sein". JSON-RPC wurde damals halt ausschließlich für die Namen benutzt.
